File: app/src/main/java/br/com/fclug/financialaid/RegisterActivity.java
package br.com.fclug.financialaid;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

import br.com.fclug.financialaid.server.ApiRequest;
import br.com.fclug.financialaid.server.ServerUtils;

public class RegisterActivity extends AppCompatActivity {

    private SessionManager mSession;
    private HashMap<String, String> mUserData;

    private EditText mName;
    private EditText mUsername;
    private EditText mPassword;
    private EditText mPasswordRepeat;
    private Button mRegister;
    private ImageView mFeedbackImage;
    private AlertDialog mAlertDialog;

    private TextWatcher mWatcher = new TextWatcher() {
        @Override
        public void beforeTextChanged(CharSequence s, int start, int count, int after) {

        }

        @Override
        public void onTextChanged(CharSequence s, int start, int before, int count) {
            String username = mUsername.getText().toString().trim();
            boolean usernameOk = false;
            if (!username.isEmpty()) {
                try {
                    ApiRequest.RequestCallback callback = new ApiRequest.RequestCallback() {
                        @Override
                        public void onSuccess(JSONObject response) throws JSONException {
                            boolean exists = response.getJSONArray("result").getJSONObject(0).getInt("existing") == 1;
                            if (exists) {
                                mFeedbackImage.setImageResource(android.R.drawable.ic_delete);
                            } else {
                                Drawable confirm = ContextCompat.getDrawable(getApplicationContext(), R.drawable.ic_confirm);
                                confirm.setColorFilter(ContextCompat.getColor(getApplicationContext(), R.color.transaction_type_credit), PorterDuff.Mode.SRC_ATOP);
                                mFeedbackImage.setImageDrawable(confirm);
                            }
                            mFeedbackImage.setVisibility(View.VISIBLE);
                        }

                        @Override
                        public void onFailure(int code) {

                        }
                    };

                    JSONObject args = new JSONObject();
                    args.put("username", username);
                    Log.d("REgister", args.toString());

                    new ApiRequest(ServerUtils.METHOD_POST, ServerUtils.ROUTE_CHECK_USERNAME, args, callback).execute();
                } catch (JSONException e) {
                    e.printStackTrace();
                }
            } else {
                mFeedbackImage.setVisibility(View.GONE);
            }
            mRegister.setEnabled(!mName.getText().toString().trim().isEmpty() &&
                    !username.isEmpty() &&
                    !mPassword.getText().toString().trim().isEmpty() &&
                    !mPasswordRepeat.getText().toString().trim().isEmpty());
        }

        @Override
        public void afterTextChanged(Editable s) {

        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.register_activity);

        mSession = new SessionManager(RegisterActivity.this);
        mUserData = mSession.getUserDetails();

        mName = (EditText) findViewById(R.id.register_name);
        mUsername = (EditText) findViewById(R.id.register_username);
        mPassword = (EditText) findViewById(R.id.register_password);
        mPasswordRepeat = (EditText) findViewById(R.id.register_password_repeat);
        mRegister = (Button) findViewById(R.id.register_button);
        mFeedbackImage = (ImageView) findViewById(R.id.register_feedback);

        mName.addTextChangedListener(mWatcher);
        mUsername.addTextChangedListener(mWatcher);
        mPassword.addTextChangedListener(mWatcher);
        mPasswordRepeat.addTextChangedListener(mWatcher);

        final ProgressDialog progressDialog = new ProgressDialog(RegisterActivity.this);
        progressDialog.setTitle("Registering");
        progressDialog.setMessage("Communicating with server...");

        mAlertDialog = new AlertDialog.Builder(RegisterActivity.this).create();
        mAlertDialog.setTitle("Registration failed!");
        mAlertDialog.setButton(DialogInterface.BUTTON_POSITIVE, "OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                mUsername.getText().clear();
                mPassword.getText().clear();
                mPasswordRepeat.getText().clear();
            }
        });

        mRegister.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                progressDialog.show();
                final String username = mUsername.getText().toString();
                final String password = mPassword.getText().toString();
                String passwordRepeat = mPasswordRepeat.getText().toString();
                if (password.equals(passwordRepeat)) {
                    JSONObject args = new JSONObject();
                    try {
                        args.put("name", mName.getText().toString());
                        args.put("username", username);
                        args.put("password", password);
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                    ApiRequest registerRequest = new ApiRequest(ServerUtils.METHOD_POST, ServerUtils.ROUTE_REGISTER_USER, args,
                            new ApiRequest.RequestCallback() {
                                @Override
                                public void onSuccess(JSONObject response) throws JSONException {
                                    progressDialog.dismiss();
                                    String token = response.getString("token");
                                    mSession.createLoginSession(username, mName.getText().toString(), password, token);
                                    Intent i = new Intent(RegisterActivity.this, MainActivity.class);
                                    startActivity(i);
                                    finish();
                                }

                                @Override
                                public void onFailure(int code) {
                                    progressDialog.dismiss();
                                }
                            });
                    registerRequest.execute();
                } else {
                    mAlertDialog.setMessage("Passwords don't match!");
                    mAlertDialog.show();
                }
            }
        });
    }
}

File: app/src/main/java/br/com/fclug/financialaid/RegisterActivity.kt
package br.com.fclug.financialaid

import android.content.DialogInterface
import android.graphics.PorterDuff
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import br.com.fclug.financialaid.server.*
import br.com.fclug.financialaid.utils.TAG
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.register_activity.*
import org.koin.android.ext.android.inject
import retrofit2.HttpException

class RegisterActivity : AppCompatActivity() {

    private var disposable: Disposable? = null
    private val api: ServerApi by inject()

    private val usernameWatcher: TextWatcher = object : TextWatcher {
        override fun afterTextChanged(s: Editable?) {}

        override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            if(s.toString().isNotEmpty()) {
                disposable = api.checkUsername(s.toString())
                        .subscribeOn(Schedulers.io())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribe({ response ->
                            val exists = response.result.existing == 1
                            if (exists) {
                                register_feedback.setImageResource(android.R.drawable.ic_delete)
                            } else {
                                ContextCompat.getDrawable(applicationContext,
                                        R.drawable.ic_confirm)?.apply {
                                    this.setColorFilter(ContextCompat.getColor(applicationContext,
                                            R.color.transaction_type_credit), PorterDuff.Mode.SRC_ATOP)
                                    register_feedback.setImageDrawable(this)
                                }
                            }
                            register_feedback.visibility = View.VISIBLE
                        }, { error ->
                            Toast.makeText(this@RegisterActivity, R.string.api_request_failed,
                                    Toast.LENGTH_SHORT).show()
                            Log.d(TAG, error.message)
                            error.printStackTrace()
                        })
            } else {
                register_feedback.visibility = View.GONE
            }
        }
    }

    private val registerEnableWatcher: TextWatcher = object : TextWatcher {
        override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}

        override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
            val username = register_username.text.toString().trim { it <= ' ' }
            val name = register_name.text.toString().trim { it <= ' ' }
            val password = register_password.text.toString()
            val passwordRepeat = register_password_repeat.text.toString()

            val passwordInputted = password.isNotEmpty() && passwordRepeat.isNotEmpty()
            val passwordValid = password == passwordRepeat

            register_password_alert.visibility = if(passwordInputted && !passwordValid)
                View.VISIBLE else View.GONE


            //TODO: add minimum size and complexity rules
            register_button.isEnabled = name.isNotEmpty() && username.isNotEmpty() &&
                    passwordInputted && passwordValid
        }

        override fun afterTextChanged(s: Editable) {}
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.register_activity)

        register_name.addTextChangedListener(registerEnableWatcher)
        register_username.addTextChangedListener(registerEnableWatcher)
        register_username.addTextChangedListener(usernameWatcher)
        register_password.addTextChangedListener(registerEnableWatcher)
        register_password_repeat.addTextChangedListener(registerEnableWatcher)

        val progressDialog = AlertDialog.Builder(this)
                .setView(R.layout.progress_dialog)
                .setTitle(resources.getString(R.string.register_dialog_title))
                .setCancelable(false)
                .create()

        val alertDialog = AlertDialog.Builder(this)
                .setTitle(resources.getString(R.string.registration_failed))
                .create()
        alertDialog.setButton(DialogInterface.BUTTON_POSITIVE, "OK") { _, _ ->
            register_username.text.clear()
            register_password.text.clear()
            register_password_repeat.text.clear()
        }
        register_button.setOnClickListener {
            val username = register_username.text.toString().trim { it <= ' ' }
            val name = register_name.text.toString().trim { it <= ' ' }
            val password = register_password.text.toString()
            val userCreation = UserCreation(username, name, password)

            progressDialog.show()
            disposable = api.register(userCreation)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({response ->
                        progressDialog.dismiss()
                        Log.d(TAG, response.result.toString())
                        val sessionManager : SessionManager by inject()
                        sessionManager.checkLogin()
                    }, {error ->
                        progressDialog.dismiss()

                        if(error is HttpException) {
                            error.response()?.errorBody()?.let {
                                val errorResponse = api.convertErrorResponseBody<ApiResponseError>(it)
                                errorResponse?.let { response ->
                                    alertDialog.setMessage(response.errors[0].description)
                                    alertDialog.show()
                                }
                            }
                        } else {
                            Toast.makeText(this@RegisterActivity, R.string.api_request_failed,
                                    Toast.LENGTH_SHORT).show()
                            Log.d(TAG, error.message)
                            error.printStackTrace()
                        }
                    })
        }
    }

    override fun onPause() {
        super.onPause()
        disposable?.dispose()
    }
}